Get list of required mount points for mount point assignment from backend

vojtechtrefny opened the pull request:
Follow up for #5066 I've added a new backend function which uses [modules/storage/platform.py](https://github.com/rhinstaller/anaconda/blob/master/pyanaconda/modules/storage/platform.py) to get the list of required boot-related required mount points.
I opening this as a draft to be able to discuss the backend changes, I plan to make some more changes in the webUI with the additional information we cna get from the platform module (like checking for correct filesystem and device types for the mount points).

VladimirSlavik commented on pyanaconda/modules/common/structures/storage.py:
```diff
@@ -453,3 +453,61 @@ def get_root_device(self):
+    """Platform required mount points data"""
+
+    def __init__(self):
+        self._value = ""
+        self._filesystem = ""
+        self._encrypted = False
+        self._lv = False
```
My first thought was that these could have actually descriptive names. Something along these lines...
- value = mountpoint, path
- encrypted = can_be_encrypted, encryption_allowed
- lv = can_be_on_lv, allowed_on_lvm...

eg. "lv" feels to me like "this is lv", not "this can be on lv"...

I see that these match `PartSpec`, so it's clear where this came from. However, for a D-Bus consumer, they will see first the names, so it might be good to make them less cryptic?

VladimirSlavik commented on ui/webui/src/components/AnacondaWizard.jsx:
```diff
@@ -60,11 +61,21 @@ export const AnacondaWizard = ({ dispatch, isBootIso, osRelease, storageData, lo
         return Object.keys(storageData.devices);
     }, [storageData.devices]);
 
+    useEffect(() => {
+        const updateRequiredMountPoints = async () => {
+            const requiredMountPoints = await getRequiredMountPoints().catch(console.error);
+
+            setRequiredMountPoints(requiredMountPoints);
+        };
+        updateRequiredMountPoints();
+    }, []);
```
I actually wonder, can't you do something like this?
```
    useEffect(() => {
            getRequiredMountPoints()
                .then(required_mount_points => setRequiredMountPoints(required_mount_points))
                .catch(console.error);
    }, []);
```

KKoukiou replied:
That's possible. It's just another style of writing this.

VladimirSlavik commented on pyanaconda/modules/common/structures/storage.py:
```diff
@@ -453,3 +453,61 @@ def get_root_device(self):
         return self.mount_points.get("/")
+
+
+class PlatformMountPointData(DBusData):
+    """Platform required mount points data"""
```
- Maybe add the Required to the name? I think this naming and docstring does not do a good job of explaining that the data is about something that is hypothetical and needs to be created, instead of something already existing.
- I wonder why have platform there? Is this related to some platform module, object, class,...?

poncovka commented on pyanaconda/modules/common/structures/storage.py:
```diff
@@ -453,3 +453,61 @@ def get_root_device(self):
+        self._lv = False
+
+    @property
+    def value(self) -> Str:
```
It would be good to keep the DBus structures consistent. So `value` should be probably `mount_point` and `fstype` should be `file_system_type`....

VladimirSlavik commented on ui/webui/src/components/storage/MountPointMapping.jsx:
```diff
@@ -130,18 +130,20 @@ const isReformatInvalid = (deviceData, request, requests) => {
-const isDeviceMountPointInvalid = (deviceData, request) => {
+const isDeviceMountPointInvalid = (deviceData, requiredMountPoints, request) => {
     const device = request["device-spec"];
+    const requiredMountPointData = requiredMountPoints.find(val => val.value.v === request["mount-point"]);
 
     if (!device || !request["mount-point"]) {
         return [false, ""];
     }
```
I'm not 100% sure, but I think you also want to guard against no matching required mount point? It seems to me you'd get a "null dereference" in the code below (whatever it's called in javascript).

VladimirSlavik reviewed (commented):
Thank you, for the backend structure, LGTM.

I don't want to pressure you into renaming all the things, but we don't want to propagate legacy anaconda internals names as API.

poncovka commented on pyanaconda/modules/common/structures/storage.py:
```diff
@@ -453,3 +453,61 @@ def get_root_device(self):
+        self._value = value
+
+    @property
+    def fstype(self) -> Str:
```
...actually, the names should be more descriptive, so something like `required_file_system_type` would make more sense.

poncovka commented on pyanaconda/modules/common/structures/storage.py:
```diff
@@ -453,3 +453,61 @@ def get_root_device(self):
+        self._fstype = fstype
+
+    @property
+    def encrypted(self) -> Bool:
```
`encryption_allowed`?

poncovka commented on pyanaconda/modules/common/structures/storage.py:
```diff
@@ -453,3 +453,61 @@ def get_root_device(self):
+        self._encrypted = encrypted
+
+    @property
+    def lv(self) -> Bool:
```
`logical_volume_allowed?`

poncovka commented on pyanaconda/modules/common/structures/storage.py:
```diff
@@ -453,3 +453,61 @@ def get_root_device(self):
         return self.mount_points.get("/")
+
+
+class PlatformMountPointData(DBusData):
```
The same structure can eventually used to expose the autopart specification. Please, take it into account when choosing a name of this structure.

vojtechtrefny replied:
I am not sure what would be the best name for this, I changed it to `RequiredMountPointData` and added a better docstring (or at least I hope it is better).

poncovka reviewed (commented):
I am not a fan of the `PartSpec` class and it is definitely a candidate for refactorization, but there was never a good time to do that. So instead of replicating the `PartSpec` class in the DBus API, it would be better to modernize the DBus structure and eventually apply these changes to the `PartSpec` class (or replace it completely).

poncovka commented on tests/unit_tests/pyanaconda_tests/modules/storage/test_module_device_tree.py:
```diff
@@ -849,6 +849,24 @@ def test_set_device_mount_options(self):
+        # we are always adding / so it's a good candidate for testing
+        root = next(r for r in result if r["value"] == get_variant(Str, "/"))
+        assert root is not None
+        assert root["value"] == get_variant(Str, "/")
```
Please, consider to use `get_native(root)` to check the content of the structure. The current test will not catch newly added fields in the future.

pep8speaks commented:
Hello @vojtechtrefny! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pyanaconda/modules/storage/devicetree/viewer.py`](https://github.com/rhinstaller/anaconda/blob/d5ce9b4619db4bd0097f01ef2df14b3cd5394e87/pyanaconda/modules/storage/devicetree/viewer.py):

> [Line 497:25](https://github.com/rhinstaller/anaconda/blob/d5ce9b4619db4bd0097f01ef2df14b3cd5394e87/pyanaconda/modules/storage/devicetree/viewer.py#L497): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent

* In the file [`pyanaconda/modules/storage/devicetree/viewer_interface.py`](https://github.com/rhinstaller/anaconda/blob/d5ce9b4619db4bd0097f01ef2df14b3cd5394e87/pyanaconda/modules/storage/devicetree/viewer_interface.py):

> [Line 204:100](https://github.com/rhinstaller/anaconda/blob/d5ce9b4619db4bd0097f01ef2df14b3cd5394e87/pyanaconda/modules/storage/devicetree/viewer_interface.py#L204): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 99 characters)

vojtechtrefny commented:
I hope I addressed all the review comments so I think this is ready for a final review. Thank you for all the suggestions for better function/argument names.

VladimirSlavik commented on tests/unit_tests/pyanaconda_tests/modules/storage/test_module_device_tree.py:
```diff
@@ -849,6 +849,26 @@ def test_set_device_mount_options(self):
 
+    def test_get_required_mount_points(self):
+        """Test GetRequiredMountPoints."""
+        result = self.interface.GetRequiredMountPoints()
+        assert isinstance(result, list)
+        assert len(result) != 0
+
+        for mp in result:
+            assert mp["mount-point"] is not None
+            assert mp["required-filesystem-type"] is not None
```
You are testing directly the type transported on D-Bus, while it's possible to convert back to the structure list. I am not 100% sure which is correct when, but I'd expect the former as it's less "raw".

Here's a few lines to show how that would look like:
```
result = RequiredMountPoint.from_structure_list(self.interface.GetRequiredMountPoints())
for mp in result:
    assert mp.mount_point is not None
root = next(r for r in result if r.mount_point == "/")
```

Your thing works, just... it's writing the python test as if it was the javascript frontend.

vojtechtrefny replied:
Ok, fixed.

VladimirSlavik commented on ui/webui/src/components/storage/MountPointMapping.jsx:
```diff
@@ -73,18 +73,20 @@ const getInitialRequests = (usablePartitioningRequests, requiredMountPoints) =>
     const rootOriginalRequest = usablePartitioningRequests.find(r => r["mount-point"] === "/");
     const bootEfiOriginalRequest = usablePartitioningRequests.find(r => r["mount-point"] === "/boot/efi");
 
+    console.log(requiredMountPoints);
```
I think you want to log also some string to explain what this is about. Otherwise it's useless clutter in the log.

vojtechtrefny replied:
Oops, that wasn't supposed to stay there. I guess I need some script to check for forgotten debug prints :-)

VladimirSlavik reviewed (approved):
I think this is fine.

vojtechtrefny commented:
/kickstart-test --testtype smoke
